src/System.Linq/tests/SelectTests.cs

```diff
@@ -0,0 +1,662 @@
+        {
+            bool funcCalled = false;
+            ICollection<Func<int>> source = new LinkedList<Func<int>>(new List<Func<int>>() { () => { funcCalled = true; return 1; } });
+
+            IEnumerable<int> query = source.Select(d => d).Select(d => d());
+            Assert.False(funcCalled);
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsIEnumerable_ExecutionIsDefered()
+        {
+            bool funcCalled = false;
+            IEnumerable<Func<int>> source = Enumerable.Repeat((Func<int>)(() => { funcCalled = true; return 1; }), 1);
+
+            IEnumerable<int> query = source.Select(d => d).Select(d => d());
+            Assert.False(funcCalled);
+        }
+
+        #endregion
+
+        #region Expected return value
+
+        [Fact]
+        public void Select_SourceIsAnArray_ReturnsExpectedValues()
+        {
+            int[] source = new[] { 1, 2, 3, 4, 5 };
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                var expected = selector(source[index]);
+                Assert.Equal(expected, item);
+                index++;
+            }
+
+            Assert.Equal(source.Length, index);
+        }
+
+        [Fact]
+        public void Select_SourceIsAList_ReturnsExpectedValues()
+        {
+            List<int> source = new List<int> { 1, 2, 3, 4, 5 };
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                var expected = selector(source[index]);
+                Assert.Equal(expected, item);
+                index++;
+            }
+
+            Assert.Equal(source.Count, index);
+        }
+
+        [Fact]
+        public void Select_SourceIsIReadOnlyCollection_ReturnsExpectedValues()
+        {
+            IReadOnlyCollection<int> source = new ReadOnlyCollection<int>(new List<int> { 1, 2, 3, 4, 5 });
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                index++;
+                var expected = selector(index);
+                Assert.Equal(expected, item);
+            }
+
+            Assert.Equal(source.Count, index);
+        }
+
+        [Fact]
+        public void Select_SourceIsICollection_ReturnsExpectedValues()
+        {
+            ICollection<int> source = new LinkedList<int>(new List<int> { 1, 2, 3, 4, 5 });
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                index++;
+                var expected = selector(index);
```

Comment:
I personally like relying on the values in `source` rather than on the index so someone could safely make modifications to the array without changing the loop structure.